source/read-write-concern/read-write-concern.rst

```diff
@@ -96,6 +96,13 @@ The read concern option is available for the following operations:
 
 ``aggregate`` command with $out and ``mapReduce`` command with "out" set to anything other than "inline" do not support readConcern.
 
+Starting in MongoDB 4.2, an ``aggregate`` command with ``$out`` supports a ``readConcern``; however, it does not support
+the "linearizable" level (attempting to do so will result in a server error). Server versions before 4.2 do not support
+a ``readConcern`` at all for ``aggregate`` commands with ``$out``.
```

Comment:
I noticed that "snapshot" wasn't mentioned or tested. Do we know if that is supported if the `aggregate` is executed within a transaction? Since the transaction's read concern is specified on the initial transaction command, I assume a test would need to use multiple operations:

 * Start transaction with RC level "snapshot"
 * Execute `aggregate` with `$out` and _do not_ provide a RC option
 * Expect `aggregate` to experience an error
 * Observe that outgoing `aggregate` command included RC "snapshot", `startTransaction`, etc.

Note: I'm assuming that "snapshot" is not supported since the motivator for this spec change was [SERVER-20445](https://jira.mongodb.org/browse/SERVER-20445), which introduced "majority" read concern support. Before then, only "local" and "available" were supported. Therefore, I would still expect other RC levels to be unsupported.

Reply:
`$out` is not supported within a transaction so you'll get an error but it won't be caused by the read concern.

Reply:
On MongoDB 4.1.8 here is the error generated: 
```
repl0:PRIMARY> db.testaggcollection.aggregate([{$sort: {x: 1}}, {$match: {_id: {$gt: 1}}}, {$out: "other_test_collection"}], {readConcern: {level: "snapshot"}})
2019-02-13T11:35:31.027-0800 E QUERY    [js] Error: command failed: {
	"operationTime" : Timestamp(1550086525, 1),
	"ok" : 0,
	"errmsg" : "readConcern level snapshot is only valid for the first transaction operation",
	"code" : 72,
	"codeName" : "InvalidOptions",
	"$clusterTime" : {
		"clusterTime" : Timestamp(1550086525, 1),
		"signature" : {
			"hash" : BinData(0,"AAAAAAAAAAAAAAAAAAAAAAAAAAA="),
			"keyId" : NumberLong(0)
		}
	}
} : aggregate failed :
_getErrorWithCode@src/mongo/shell/utils.js:25:13
doassert@src/mongo/shell/assert.js:18:14
_assertCommandWorked@src/mongo/shell/assert.js:584:17
assert.commandWorked@src/mongo/shell/assert.js:674:16
DB.prototype._runAggregate@src/mongo/shell/db.js:260:9
DBCollection.prototype.aggregate@src/mongo/shell/collection.js:987:12
@(shell):1:1
```

@jmikola do you think I should explicitly mention that "snapshot" `readConcern` cannot be used with an aggregation pipeline containing the `$out` stage due to the limitation with transactions that Shane mentioned?

Reply:
> `$out` is not supported within a transaction so you'll get an error but it won't be caused by the read concern.

Ah, didn't realize that. Thanks!

> do you think I should explicitly mention that "snapshot" `readConcern` cannot be used with an aggregation pipeline containing the `$out` stage due to the limitation with transactions that Shane mentioned?

Rather than attempt to document what _isn't_ supported, I think it's easier to document what _is_ supported.

I think not originally doing so is how we ended up with the original spec language that this PR is changing. Previously, the spec said that `$out` did not support _any_ read concern. This might have been an over-simplification, because  "local" is essentially the default behavior and "available" (introduced a bit later in 3.6) behaves the same as "local" for unsharded collections. At the time, "majority" was probably the only read concern level with a functional difference and since it was not supported, the spec wrote off all read concerns.

Looking at your tests, you expect "local", "available", and "majority" to succeed, so I propose we explicitly list supported read concerns. This will help future-proof the spec, since we don't know what may be added down the line. We can always add a new supported level at a later time.

Also, saying that "snapshot" is unsupported might be accurate but a bit odd, since we know it simply cannot be used by virtue of `$out` not supporting transactions (and "snapshot" only relates to transactions). I think that's another reason why it'd preferable to just list the levels we know _are_ supported.

Reply:
Thanks for laying that out so nicely! I'll modify the text here to explicitly mention the supported readConcerns.